File: apps/ghost-tap/ios/GhostTap/ViewModels/WalletViewModel.swift
import Foundation
import Combine

/// Sync status for the wallet
enum SyncStatus: Equatable {
    case idle
    case syncing(progress: Double)
    case synced(height: UInt64)
    case error(String)

    var description: String {
        switch self {
        case .idle: return "Idle"
        case .syncing(let p): return "Syncing \(Int(p * 100))%"
        case .synced(let h): return "Synced (block \(h))"
        case .error(let e): return "Error: \(e)"
        }
    }
}

/// Fee priority levels matching Rust FeePriority enum (0=Low, 1=Med, 2=High)
enum FeePriority: UInt8, CaseIterable, Identifiable {
    case low = 0
    case medium = 1
    case high = 2

    var id: UInt8 { rawValue }

    var label: String {
        switch self {
        case .low: return "Low"
        case .medium: return "Medium"
        case .high: return "High"
        }
    }

    var description: String {
        switch self {
        case .low: return "~60 min"
        case .medium: return "~20 min"
        case .high: return "~5 min"
        }
    }
}

/// A simplified transaction model for the UI layer
struct Transaction: Identifiable {
    let id: String          // txid
    let direction: String   // "incoming" or "outgoing"
    let amount: UInt64
    let fee: UInt64?
    let address: String
    let status: String      // "pending", "confirmed", "failed"
    let confirmations: UInt32
    let timestamp: UInt64
    let memo: String?

    var isIncoming: Bool { direction == "incoming" }

    /// Human-readable relative timestamp
    var relativeTime: String {
        let date = Date(timeIntervalSince1970: TimeInterval(timestamp))
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: date, relativeTo: Date())
    }

    /// Truncated address for display: first 8 ... last 6
    var addressSnippet: String {
        guard address.count > 16 else { return address }
        let prefix = address.prefix(8)
        let suffix = address.suffix(6)
        return "\(prefix)...\(suffix)"
    }
}

/// Main wallet view model wrapping the Rust WalletHandle via UniFFI
@MainActor
final class WalletViewModel: ObservableObject {

    init() {
        let stored = UserDefaults.standard.integer(forKey: "autoLockMinutes")
        self.autoLockMinutes = stored > 0 ? stored : 5
    }

    // MARK: - Published State

    @Published var balance: UInt64 = 0
    @Published var pendingIncoming: UInt64 = 0
    @Published var pendingOutgoing: UInt64 = 0
    @Published var history: [Transaction] = []
    @Published var addresses: [String] = []
    @Published var currentAddress: String = ""
    @Published var isLocked: Bool = false
    @Published var syncStatus: SyncStatus = .idle
    @Published var hasWallet: Bool = false
    @Published var mnemonic: [String] = []
    @Published var errorMessage: String?

    // MARK: - Auto-Lock

    /// Minutes of inactivity before the wallet locks (user-configurable).
    @Published var autoLockMinutes: Int {
        didSet {
            UserDefaults.standard.set(autoLockMinutes, forKey: "autoLockMinutes")
        }
    }

    /// Timestamp of the last user activity (set when app enters background).
    private var lastActivityTimestamp: Date?

    /// Record that the user is active (call when app enters background).
    func recordActivity() {
        lastActivityTimestamp = Date()
    }

    /// Check if the wallet should be locked based on elapsed time
    /// (call when app enters foreground).
    func checkAutoLock() {
        guard hasWallet, !isLocked, hasPin() else { return }
        guard let last = lastActivityTimestamp else { return }

        let elapsed = Date().timeIntervalSince(last)
        if elapsed >= Double(autoLockMinutes) * 60.0 {
            lockWallet()
        }
    }

    // MARK: - Internal

    private var walletHandle: WalletHandle?

    // MARK: - Wallet Lifecycle

    /// Create a new wallet, storing mnemonic words for backup flow
    func createWallet(wordCount: Int = 12) async {
        do {
            let handle: WalletHandle
            if wordCount == 24 {
                handle = try WalletHandle.generate24()
            } else {
                handle = try WalletHandle.generate12()
            }

            walletHandle = handle
            mnemonic = handle.getMnemonic().split(separator: " ").map(String.init)
            hasWallet = true
            isLocked = false

            await generateAddress()
            await refreshBalance()
        } catch {
            errorMessage = "Failed to create wallet: \(error.localizedDescription)"
        }
    }

    /// Import a wallet from a mnemonic phrase
    func importWallet(phrase: String, passphrase: String? = nil) async -> Bool {
        guard walletValidateMnemonic(mnemonic: phrase) else {
            errorMessage = "Invalid mnemonic phrase"
            return false
        }

        do {
            let handle = try WalletHandle.fromMnemonic(
                mnemonic: phrase,
                passphrase: passphrase
            )
            walletHandle = handle
            mnemonic = phrase.split(separator: " ").map(String.init)
            hasWallet = true
            isLocked = false

            await generateAddress()
            await refreshBalance()
            return true
        } catch {
            errorMessage = "Failed to import wallet: \(error.localizedDescription)"
            return false
        }
    }

    // MARK: - Balance

    func refreshBalance() async {
        guard let handle = walletHandle else { return }

        do {
            let details = try handle.getBalanceDetails()
            balance = details.confirmed
            pendingIncoming = details.pendingIncoming
            pendingOutgoing = details.pendingOutgoing
        } catch {
            balance = handle.getBalance()
        }

        await refreshHistory()
    }

    // MARK: - Addresses

    func generateAddress() async {
        guard let handle = walletHandle else { return }
        do {
            let addr = try handle.newReceiveAddress()
            currentAddress = addr
            if !addresses.contains(addr) {
                addresses.insert(addr, at: 0)
            }
        } catch {
            errorMessage = "Failed to generate address: \(error.localizedDescription)"
        }
    }

    // MARK: - History

    func refreshHistory() async {
        guard let handle = walletHandle else { return }
        do {
            let entries = try handle.getHistory(offset: 0, limit: 100)
            history = entries.map { entry in
                Transaction(
                    id: entry.txid,
                    direction: entry.direction,
                    amount: entry.amount,
                    fee: entry.fee,
                    address: entry.address,
                    status: entry.status,
                    confirmations: entry.confirmations,
                    timestamp: entry.timestamp,
                    memo: entry.memo
                )
            }
        } catch {
            errorMessage = "Failed to load history: \(error.localizedDescription)"
        }
    }

    // MARK: - Send

    /// Build and broadcast a transaction. Returns the txid on success.
    func send(to address: String, amount: UInt64, fee: FeePriority) async -> String? {
        guard let handle = walletHandle else {
            errorMessage = "No wallet loaded"
            return nil
        }

        do {
            let unsigned = try handle.buildTransaction(
                toAddress: address,
                amount: amount,
                feePriority: fee.rawValue
            )
            let txid = try handle.signAndBroadcast(unsignedTxJson: unsigned.txJson)
            await refreshBalance()
            return txid
        } catch {
            errorMessage = "Send failed: \(error.localizedDescription)"
            return nil
        }
    }

    // MARK: - Lock

    func lockWallet() {
        walletHandle?.lock()
        isLocked = true
    }

    func unlockWallet() {
        walletHandle?.unlock()
        isLocked = false
    }

    // MARK: - Formatting Helpers

    static func formatBalance(_ satoshis: UInt64) -> String {
        let whole = satoshis / 100_000_000
        let frac = satoshis % 100_000_000
        return String(format: "%d.%08d", whole, frac)
    }
}

File: apps/ghost-tap/android/app/src/main/kotlin/com/ghost/tap/viewmodel/WalletViewModel.kt
package com.ghost.tap.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ghost.tap.RustBridge
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

data class Transaction(
    val txId: String,
    val amount: Long,
    val fee: Long,
    val address: String,
    val timestamp: Long,
    val confirmations: Int,
    val isSend: Boolean,
    val status: TxStatus
)

enum class TxStatus { Pending, Confirmed, Failed }

enum class SyncStatus { Idle, Syncing, Synced, Error }

data class WalletUiState(
    val hasWallet: Boolean = false,
    val balance: Long = 0L,
    val transactions: List<Transaction> = emptyList(),
    val currentAddress: String = "",
    val syncStatus: SyncStatus = SyncStatus.Idle,
    val walletHandle: Long = -1L,
    val error: String? = null,
    val isGenerating: Boolean = false,
    val isSending: Boolean = false,
    val sendResult: SendResult? = null,
    val isLocked: Boolean = false
)

sealed class SendResult {
    data class Success(val txId: String) : SendResult()
    data class Error(val message: String) : SendResult()
}

class WalletViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(WalletUiState())
    val uiState: StateFlow<WalletUiState> = _uiState.asStateFlow()

    /** Minutes of inactivity before auto-lock (default 5). */
    var autoLockMinutes: Int = 5

    private var lastActivityTimestamp: Long = 0L

    /** Record activity timestamp (call from onPause). */
    fun recordActivity() {
        lastActivityTimestamp = System.currentTimeMillis()
    }

    /** Check if wallet should auto-lock (call from onResume). */
    fun checkAutoLock() {
        if (!_uiState.value.hasWallet || _uiState.value.isLocked) return
        if (lastActivityTimestamp == 0L) return

        val elapsed = System.currentTimeMillis() - lastActivityTimestamp
        if (elapsed >= autoLockMinutes * 60_000L) {
            lockWallet()
        }
    }

    fun createWallet(use24Words: Boolean = false): MutableStateFlow<String?> {
        val mnemonicFlow = MutableStateFlow<String?>(null)
        _uiState.update { it.copy(isGenerating = true, error = null) }

        viewModelScope.launch {
            val mnemonic = withContext(Dispatchers.IO) {
                if (use24Words) RustBridge.generateWallet24() else RustBridge.generateWallet12()
            }
            if (mnemonic != null) {
                mnemonicFlow.value = mnemonic
            } else {
                _uiState.update { it.copy(error = "Failed to generate wallet") }
            }
            _uiState.update { it.copy(isGenerating = false) }
        }
        return mnemonicFlow
    }

    fun finalizeWalletCreation(mnemonic: String) {
        viewModelScope.launch {
            val handle = withContext(Dispatchers.IO) {
                RustBridge.importWallet(mnemonic, null)
            }
            if (handle >= 0) {
                _uiState.update {
                    it.copy(hasWallet = true, walletHandle = handle, error = null)
                }
                generateAddress()
                refreshBalance()
            } else {
                _uiState.update { it.copy(error = "Failed to initialize wallet") }
            }
        }
    }

    fun importWallet(mnemonic: String, passphrase: String? = null) {
        _uiState.update { it.copy(isGenerating = true, error = null) }

        viewModelScope.launch {
            val trimmed = mnemonic.trim().lowercase()
            val valid = withContext(Dispatchers.IO) {
                RustBridge.validateMnemonic(trimmed)
            }
            if (!valid) {
                _uiState.update {
                    it.copy(isGenerating = false, error = "Invalid mnemonic phrase")
                }
                return@launch
            }

            val handle = withContext(Dispatchers.IO) {
                RustBridge.importWallet(trimmed, passphrase)
            }
            if (handle >= 0) {
                _uiState.update {
                    it.copy(
                        hasWallet = true,
                        walletHandle = handle,
                        isGenerating = false,
                        error = null
                    )
                }
                generateAddress()
                refreshBalance()
            } else {
                _uiState.update {
                    it.copy(isGenerating = false, error = "Failed to import wallet")
                }
            }
        }
    }

    fun refreshBalance() {
        val handle = _uiState.value.walletHandle
        if (handle < 0) return

        _uiState.update { it.copy(syncStatus = SyncStatus.Syncing) }

        viewModelScope.launch {
            try {
                val balance = withContext(Dispatchers.IO) {
                    RustBridge.getBalance(handle)
                }
                val historyJson = withContext(Dispatchers.IO) {
                    RustBridge.getTransactionHistory(handle)
                }
                val transactions = parseTransactions(historyJson)

                _uiState.update {
                    it.copy(
                        balance = balance,
                        transactions = transactions,
                        syncStatus = SyncStatus.Synced,
                        error = null
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(syncStatus = SyncStatus.Error, error = e.message)
                }
            }
        }
    }

    fun generateAddress() {
        val handle = _uiState.value.walletHandle
        if (handle < 0) return

        viewModelScope.launch {
            val address = withContext(Dispatchers.IO) {
                RustBridge.newReceiveAddress(handle)
            }
            if (address != null) {
                _uiState.update { it.copy(currentAddress = address) }
            }
        }
    }

    fun sendPayment(toAddress: String, amount: Long, feePriority: Int) {
        val handle = _uiState.value.walletHandle
        if (handle < 0) return

        _uiState.update { it.copy(isSending = true, sendResult = null) }

        viewModelScope.launch {
            try {
                val rawTx = withContext(Dispatchers.IO) {
                    RustBridge.createTransaction(handle, toAddress, amount, feePriority)
                }
                if (rawTx == null) {
                    _uiState.update {
                        it.copy(
                            isSending = false,
                            sendResult = SendResult.Error("Failed to create transaction")
                        )
                    }
                    return@launch
                }

                val txId = withContext(Dispatchers.IO) {
                    RustBridge.broadcastTransaction(rawTx)
                }
                if (txId != null) {
                    _uiState.update {
                        it.copy(isSending = false, sendResult = SendResult.Success(txId))
                    }
                    refreshBalance()
                } else {
                    _uiState.update {
                        it.copy(
                            isSending = false,
                            sendResult = SendResult.Error("Failed to broadcast transaction")
                        )
                    }
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSending = false,
                        sendResult = SendResult.Error(e.message ?: "Unknown error")
                    )
                }
            }
        }
    }

    fun clearSendResult() {
        _uiState.update { it.copy(sendResult = null) }
    }

    fun lockWallet() {
        _uiState.update { it.copy(isLocked = true) }
    }

    fun unlockWallet() {
        _uiState.update { it.copy(isLocked = false) }
    }

    fun getTransaction(txId: String): Transaction? {
        return _uiState.value.transactions.find { it.txId == txId }
    }

    private fun parseTransactions(json: String): List<Transaction> {
        return try {
            val array = JSONArray(json)
            (0 until array.length()).map { i ->
                val obj = array.getJSONObject(i)
                Transaction(
                    txId = obj.getString("txid"),
                    amount = obj.getLong("amount"),
                    fee = obj.optLong("fee", 0),
                    address = obj.optString("address", ""),
                    timestamp = obj.optLong("timestamp", 0),
                    confirmations = obj.optInt("confirmations", 0),
                    isSend = obj.optBoolean("is_send", false),
                    status = when (obj.optString("status", "pending")) {
                        "confirmed" -> TxStatus.Confirmed
                        "failed" -> TxStatus.Failed
                        else -> TxStatus.Pending
                    }
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }
}
